File: src/lib/moon.test.ts
import { describe, expect, it } from "vitest";
import { getMoonTimes } from "./moon";

describe("getMoonTimes", () => {
  // Reference values from the upstream SunCalc test suite.
  it("returns the correct moon rise/set times", () => {
    const times = getMoonTimes(new Date("2013-03-04T00:00:00Z"), 50.5, 30.5, true);
    expect(times.rise?.toUTCString()).toBe("Mon, 04 Mar 2013 23:54:29 GMT");
    expect(times.set?.toUTCString()).toBe("Mon, 04 Mar 2013 07:47:58 GMT");
    expect(times.alwaysUp).toBe(false);
    expect(times.alwaysDown).toBe(false);
  });

  it("handles days where the moon does not rise", () => {
    // High-arctic winter: moon stays below the horizon around new moon.
    const times = getMoonTimes(new Date("2023-01-21T00:00:00Z"), 80, 0, true);
    expect(times.rise).toBeNull();
    expect(times.set).toBeNull();
    expect(times.alwaysDown).toBe(true);
    expect(times.alwaysUp).toBe(false);
  });

  it("handles days where the moon stays up", () => {
    // High-arctic winter around full moon: moon circles above the horizon.
    const times = getMoonTimes(new Date("2023-01-06T00:00:00Z"), 80, 0, true);
    expect(times.rise).toBeNull();
    expect(times.set).toBeNull();
    expect(times.alwaysUp).toBe(true);
    expect(times.alwaysDown).toBe(false);
  });
});

File: mobile/test/moon_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:viam_chartplotter_mobile/moon.dart';

// The web test compares Date.toUTCString(), which truncates to whole seconds;
// this reproduces that comparison granularity.
DateTime _truncToSecond(DateTime d) {
  final u = d.toUtc();
  return DateTime.utc(u.year, u.month, u.day, u.hour, u.minute, u.second);
}

void main() {
  // Reference values from the upstream SunCalc test suite.
  test('returns the correct moon rise/set times', () {
    final times =
        getMoonTimes(DateTime.utc(2013, 3, 4), 50.5, 30.5, inUTC: true);
    expect(_truncToSecond(times.rise!), DateTime.utc(2013, 3, 4, 23, 54, 29));
    expect(_truncToSecond(times.set!), DateTime.utc(2013, 3, 4, 7, 47, 58));
    expect(times.alwaysUp, isFalse);
    expect(times.alwaysDown, isFalse);
  });

  test('handles days where the moon does not rise', () {
    // High-arctic winter: moon stays below the horizon around new moon.
    final times = getMoonTimes(DateTime.utc(2023, 1, 21), 80, 0, inUTC: true);
    expect(times.rise, isNull);
    expect(times.set, isNull);
    expect(times.alwaysDown, isTrue);
    expect(times.alwaysUp, isFalse);
  });

  test('handles days where the moon stays up', () {
    // High-arctic winter around full moon: moon circles above the horizon.
    final times = getMoonTimes(DateTime.utc(2023, 1, 6), 80, 0, inUTC: true);
    expect(times.rise, isNull);
    expect(times.set, isNull);
    expect(times.alwaysUp, isTrue);
    expect(times.alwaysDown, isFalse);
  });
}
